### PR TITLE
exploding our single dockerfile into several dockerfiles in an effort to increase build speed and better caching

### PR DESCRIPTION
**Pull request by julien51**
We are getting good results on this so I think this should be our default approach.
We now build "local" images from our "master" images for every tests but use a lot of caching.
Comments in the file changes for clarity.

- [X] 1 PR, 1 purpose: my Pull Request applies to a single purpose
  - [X] This PR only contains configuration changes (package.json, etc.)
  - [ ] This PR only contains code changes (if configuration changes are required, do a separate PR first, then re-base)
- [ ] My code follows the style guidelines of this project, including naming conventions
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] If my code adds or changes components, I have written corresponding stories with Storybook
- [x] I have performed a self-review of my own code
- [ ] If my code involves visual changes, I am adding applicable screenshots to this thread

### CODE REVIEW

File: docker/docker-compose.ci.yml [hunk 1]
```diff
@@ -21,24 +21,18 @@ services:
       POSTGRES_PASSWORD: '${DB_PASSWORD}'
       POSTGRES_DB: '${DB_NAME}'
   unlock-contract-deployer: # Used for integration tests to deploy the smart contract
-    image: unlock
-    user: node
```
Comment (julien51) [inline docker/docker-compose.ci.yml]: this is inside `unlock-core` which is a "parent" image for `unlock-app`. No need to duplicate.

File: docker/docker-compose.ci.yml [hunk 2]
```diff
@@ -21,24 +21,18 @@ services:
...
-    working_dir: /home/unlock/unlock-app
```
Comment (julien51) [inline docker/docker-compose.ci.yml]: idem

File: docker/docker-compose.ci.yml [hunk 3]
```diff
@@ -83,12 +73,13 @@ services:
...
-    working_dir: /home/unlock/wedlocks
+    image: wedlocks
+    environment:
+      CI: '${CI}'
+      UNLOCK_ENV: '${UNLOCK_ENV}'
+      SMTP_HOST: '${SMTP_HOST}'
   integration-tests:
-    image: unlock-integration
-    user: node
+    image: integration-tests
```
Comment (julien51) [inline docker/docker-compose.ci.yml]: Changed the name for consistency with the rest of the app.

File: docker/integration-tests.dockerfile [hunk 4]
```diff
@@ -40,5 +40,6 @@ WORKDIR /home/unlock/
 COPY --chown=node ./scripts /home/unlock/scripts
 
 # Copy the parent binaries from the root into the children
-WORKDIR /home/unlock/
```
Comment (julien51) [inline docker/integration-tests.dockerfile]: This is the default in parent image

File: docker/locksmith.dockerfile [hunk 5]
```diff
@@ -0,0 +1,17 @@
+FROM unlock-core
```
Comment (julien51) [inline docker/locksmith.dockerfile]: new image ;)

File: docker/unlock-core.dockerfile [hunk 6]
```diff
@@ -0,0 +1,42 @@
+FROM node:8.11.4-alpine
```
Comment (julien51) [inline docker/unlock-core.dockerfile]: This new image is our "parent" image with all the stuff that is shared between all other images.

File: scripts/build-image.sh [hunk 7]
```diff
@@ -2,21 +2,30 @@
...
+SKIP_CORE=$2
 
-if [ "$DOCKER_REPOSITORY" != "" ]; then
-  IMAGE_CACHE="$DOCKER_REPOSITORY/$IMAGE_NAME:master"
-  echo "Pulling $IMAGE_CACHE to use as cache for $IMAGE_NAME"
-  docker pull $IMAGE_CACHE;
-else
-  IMAGE_CACHE="$IMAGE_NAME:master"
+if [ "$SKIP_CORE" != "true" ]; then
+  echo "Pulling $DOCKER_REPOSITORY/unlock-core:$CACHED_IMAGE_TAG to use as cache for unlock-core"
```
Comment (julien51) [inline scripts/build-image.sh]: For each image we first fetch the docker-core image

File: scripts/build-image.sh [hunk 8]
```diff
@@ -2,21 +2,30 @@
...
+  docker pull "$DOCKER_REPOSITORY/unlock-core:master" &
 fi
 
-ARGS="$ARGS --cache-from $IMAGE_CACHE"
+IMAGE_CACHE="$DOCKER_REPOSITORY/$IMAGE_NAME:$CACHED_IMAGE_TAG"
+echo "Pulling $IMAGE_CACHE to use as cache for $IMAGE_NAME"
```
Comment (julien51) [inline scripts/build-image.sh]: And build the master cache

File: wedlocks/webpack.config.js [hunk 9]
```diff
@@ -10,7 +10,9 @@ const requiredVariables = ['SMTP_HOST', 'SMTP_USERNAME', 'SMTP_PASSWORD']
 requiredVariables.forEach(envVar => {
   if (!process.env[envVar]) {
     console.error(`Environment variable ${envVar} is required.`)
-    process.exit(1)
+    if (['dev', 'test'].indexOf(process.env.UNLOCK_ENV) > -1) {
```
Comment (akeem) [inline wedlocks/webpack.config.js]: This makes sense, but I don't think it is directly related to this change

Comment (julien51) [reply wedlocks/webpack.config.js]: That is right. I could not build the unlock image without it though :/

**Pull request merged**